staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -338,6 +362,28 @@ func New(opts Options) (*Authenticator, error) {
 	return authenticator, nil
 }
 
+// discoveryURLRoundTripper is a http.RoundTripper that rewrites the issuer URL to the
+// discovery URL for the /.well-known/openid-configuration endpoint.
+type discoveryURLRoundTripper struct {
+	base         http.RoundTripper
+	discoveryURL *url.URL
+}
+
+func (t *discoveryURLRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// If the discovery URL is set in authentication configuration, we should
+	// use the discovery URL to fetch the openid configuration. This is useful
+	// for self-hosted providers, for example, providers that run on top of
+	// Kubernetes itself.
+	if req.Method == http.MethodGet && strings.HasSuffix(req.URL.Path, wellKnownEndpointPath) {
+		clone := req.Clone(req.Context())
```

Comment:
do we need to override or clear req.Host?

do we need to check the whole URL of the intercepted request to make sure it matched `url`, not just a path suffix?

Reply:
> do we need to override or clear req.Host?

We're doing a deepcopy here and setting `clone.URL.Host` to the [new host](https://github.com/kubernetes/kubernetes/pull/123527/files#diff-de8a853b396792a685ba049763f700e10c7e03bc4482f6801e857e2f9a1b9f1eR379). That overrides the host?

> do we need to check the whole URL of the intercepted request to make sure it matched url, not just a path suffix?

The `go-oidc` library only makes the request to `/.well-known/openid-configuration` once as part of the new provider initialization and in case of the authenticator it's always with the configured issuer url. If we want to be exact here, we could check the url matches `opts.JWTAuthenticator.Issuer.URL + slashIfNeeded + wellKnownEndpointPath`. LMK if we should do that?

Reply:
there's req.URL.Host and req.Host ... I don't have call paths clear enough in my head to know if the latter can be set as well

Reply:
> If we want to be exact here, we could check the url matches opts.JWTAuthenticator.Issuer.URL + slashIfNeeded + wellKnownEndpointPath. LMK if we should do that?

I would check the whole thing

Reply:
I did a test and I do see `clone.Host` being overridden to `clone.URL.Host` after it's set.

Reply:
the `req.Host` field takes precedence if set, I think:

https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/net/http/request.go;l=594-604

so maybe clear it from our clone?

Reply:
done!